src/Sarif.Multitool/Rules/SARIF1008.PhysicalLocationPropertiesMustBeConsistent.cs

```diff
@@ -18,15 +21,149 @@ public class PhysicalLocationPropertiesMustBeConsistent : SarifValidationSkimmer
+            // It is a potential false negative.
+            return true;
+        }
+
+        private static bool IsLineColumnBasedTextRegionProperSuperset(Region superRegion, Region subRegion)
+        {
+            if (superRegion.StartLine > subRegion.StartLine || superRegion.EndLine < subRegion.EndLine)
+            {
+                return false;
+            }
+
+            if (superRegion.StartLine == subRegion.StartLine && superRegion.StartColumn > subRegion.StartColumn)
+            {
+                return false;
+            }
+
+            if (superRegion.EndLine == subRegion.EndLine && superRegion.EndColumn < subRegion.EndColumn)
+            {
+                return false;
+            }
+
+            if (superRegion.StartLine == subRegion.StartLine &&
+                superRegion.EndLine == subRegion.EndLine &&
+                superRegion.StartColumn == subRegion.StartColumn &&
+                superRegion.EndColumn == subRegion.EndColumn)
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        private static bool IsBinaryRegionProperSuperset(Region superRegion, Region subRegion)
+        {
+            if (superRegion.ByteOffset > subRegion.ByteOffset)
+            {
+                return false;
+            }
+
+            if (GetByteEndOffset(superRegion) < GetByteEndOffset(subRegion))
+            {
+                return false;
+            }
+
+            if (superRegion.ByteOffset == subRegion.ByteOffset && superRegion.ByteLength <= subRegion.ByteLength)
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        private static bool IsOffsetBasedTextRegionProperSupetSet(Region superRegion, Region subRegion)
+        {
+            if (superRegion.CharOffset > subRegion.CharOffset)
+            {
+                return false;
+            }
+
+            if (GetCharEndOffset(superRegion) < GetCharEndOffset(subRegion))
+            {
+                return false;
+            }
+
+            if (superRegion.CharOffset == subRegion.CharOffset && superRegion.CharLength <= subRegion.CharLength)
+            {
+                return false;
             }
+
+            return true;
+        }
+
+        private static bool IsLineColumnBasedTextRegion(Region region)
+        {
+            return region.StartLine >= 1;
+        }
+
+        private static bool IsOffsetBasedTextRegion(Region region)
+        {
+            return region.CharOffset > 0;
+        }
+
+        private static bool IsBinaryRegion(Region region)
+        {
+            return region.ByteOffset >= 0;
+        }
+
+        private static int GetCharEndOffset(Region region)
+        {
+            return region.CharOffset + region.CharLength;
+        }
+
+        private static int GetByteEndOffset(Region region)
+        {
+            return region.ByteOffset + region.ByteLength;
         }
```

Comment:
we could move all these to the Region partial class. what do u think?

Reply:
Yes, that makes sense. We would expose only the top level API `public bool IsProperSuperSetOf(Region region)` so you'd call it as `contextRegion.IsProperSupersetOfRegion(region)`. All the others would remain private. Not needed in this PR.

---
In reply to: [443893453](https://github.com/microsoft/sarif-sdk/pull/1925#discussion_r443893453) [](ancestors = 443893453)

Reply:
i decided to create static method instead. the code is much more readable that way. If you look at the implementation. SuperRegion.IsBlah && SubRegion.IsBlah is easier to understand than, this.IsBlah && OtherRegion.IsBlah

---
In reply to: [444234139](https://github.com/microsoft/sarif-sdk/pull/1925#discussion_r444234139) [](ancestors = 444234139,443893453)

Reply:
Sorry, I don't agree with this. Remember that the goal (one of the goals, anyway) of the API is to make the _caller's_ code look nice -- not the implementers! For the caller, this:
```
if (!physicalLocation.ContextRegion.IsProperSupersetOf(physicalLocation.Region))
{
    // Report error
}
```
reads better (IMNSHO) than this:
```
if (!Region.IsProperSuperset(physicalLocation.ContextRegion, physicalLocation.Region))
{
    // Report error
}
```

Read each one aloud as an English sentence to see what I mean.

(Note my last minute change of heart on the API name: `IsProperSupersetOf`.)

---
In reply to: [444584362](https://github.com/microsoft/sarif-sdk/pull/1925#discussion_r444584362) [](ancestors = 444584362,444234139,443893453)